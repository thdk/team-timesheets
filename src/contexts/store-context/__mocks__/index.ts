export const useStore = jest.fn();
