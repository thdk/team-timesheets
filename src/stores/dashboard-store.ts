import { observable, reaction, computed, action, makeObservable } from "mobx";
import { ICollection, Collection, RealtimeMode } from "firestorable";

import moment from 'moment';

import { IRootStore } from "./root-store";
import { IGroupedRegistrations } from "./registration-store/registration-store";
import * as deserializer from '../../common/serialization/deserializer';
import * as serializer from '../../common/serialization/serializer';
import { TimePeriod } from "../components/time-period-select";
import { IRegistration, IRegistrationData } from "../../common/dist";
import { Firestore, query, Query, where } from "firebase/firestore";

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;
    readonly setTimePeriodFilter: (timePeriod: TimePeriod) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
    readonly timePeriodFilterValue?: TimePeriod;

}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration, IRegistrationData>;

    private userFilterValueField: string | undefined = undefined;
    private taskFilterValueField: string | undefined = undefined;
    private projectFilterValueField: string | undefined = undefined;
    private timePeriodFilterField: TimePeriod | undefined = undefined;

    constructor(
        rootStore: IRootStore,
        {
            firestore,
        }: {
            firestore: Firestore,
        },
    ) {
        makeObservable<DashboardStore, "userFilterValueField" | "taskFilterValueField" | "projectFilterValueField" | "timePeriodFilterField" | "startDateFilter" | "endDateFilter" | "userIdFilter" | "projectIdFilter" | "startDate" | "endDate" | "registrations">(this, {
            userFilterValueField: observable.ref,
            taskFilterValueField: observable.ref,
            projectFilterValueField: observable.ref,
            timePeriodFilterField: observable.ref,
            startDateFilter: computed,
            endDateFilter: computed,
            userIdFilter: computed,
            projectIdFilter: computed,
            registrationsGroupedByUser: computed,
            registrationsGroupedByProject: computed,
            registrationsGroupedByTask: computed,
            startDate: computed,
            endDate: computed,
            registrations: computed,
            setProjectFilter: action,
            setTaskFilter: action,
            setUserFilter: action,
            setTimePeriodFilter: action
        });

        this.registrationsField = new Collection<IRegistration, IRegistrationData>(
            firestore,
            "registrations",
            {
                realtimeMode: RealtimeMode.on,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration,
                name: "Dashboard registrations",
            },
            {
                // logger: console.log
            },
        );

        const updateRegistrationQuery = () => {
            if (!rootStore.user.divisionUser || !this.timePeriodFilterField) {
                this.registrationsField.query = null;
            } else {
                this.registrationsField.query = ref =>
                    this.userIdFilter(
                        this.projectIdFilter(
                            this.endDateFilter(
                                this.startDateFilter(
                                    query(ref, where("deleted", "==", false))
                                )
                            )
                        )
                    ) as Query<IRegistrationData>;
            }
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.timePeriodFilterField, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);
        reaction(() => rootStore.user.divisionUser, updateRegistrationQuery);
    }

    private get startDateFilter() {
        const startDate = this.startDate;
        return (q: Query) => startDate ? query(q, where("date", ">=", startDate)) : q;
    }

    private get endDateFilter() {
        const endDate = this.endDate;
        return (q: Query) => endDate ? query(q, where("date", "<=", endDate)) : q;
    }

    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (q: Query) => userId ? query(q, where("userId", "==", userId)) : q;
    }

    private get projectIdFilter() {
        return (q: Query) => this.projectFilterValue
            ? query(q, where("project", "==", this.projectFilterValue))
            : q;
    }

    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    private get startDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().startOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().startOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').startOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().startOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').startOf('year').toDate();
            default:
                return undefined;
        }
    }

    private get endDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().endOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().endOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').endOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().endOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').endOf('year').toDate();
            default:
                return undefined;
        }
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    public get timePeriodFilterValue() {
        return this.timePeriodFilterField;
    }

    private get registrations() {
        return this.registrationsField.docs
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    setTimePeriodFilter(timePeriod: TimePeriod) {
        this.timePeriodFilterField = timePeriod;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0,
                        isCollapsed: true
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}