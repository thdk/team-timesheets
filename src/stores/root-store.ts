import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { IProjectStore, ProjectStore } from "./project-store";
import { FavoriteStore } from "./favorite-store";

export interface IRootStore {
    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly favorites: FavoriteStore;
    readonly dashboard: IDashboardStore;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;
    public readonly favorites: FavoriteStore;

    constructor() {
        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);
        this.favorites = new FavoriteStore(this);
    }
};
