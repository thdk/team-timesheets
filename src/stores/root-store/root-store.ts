import { RouterStore } from "mobx-router";

import { IRegistrationsStore, RegistrationStore } from "../registration-store/registration-store";
import { IConfigStore, ConfigStore } from "../config-store";
import { IUserStore, UserStore } from "../user-store";
import { IViewStore, ViewStore } from "../view-store";
import { IReportStore, ReportStore } from "../report-store/report-store";
import { DashboardStore, IDashboardStore } from "../dashboard-store";
import { ProjectStore } from "../project-store";
import { FavoriteStore } from "../favorite-store";
import { DivisionStore  } from "../division-store/division-store";

export interface IRootStore extends Store { };

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router: RouterStore<IRootStore>;
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: ProjectStore;
    public readonly favorites: FavoriteStore;
    public readonly divisions: DivisionStore;

    constructor({
        auth,
        firestore,
        httpsCallable,
    }: {
        firestore: firebase.firestore.Firestore,
        httpsCallable?: (name: string) => firebase.functions.HttpsCallable,
        auth?: firebase.auth.Auth,
    }) {
        this.user = new UserStore(
            this,
            {
                auth,
                firestore,
            }
        );

        this.view = new ViewStore();
        this.config = new ConfigStore(
            this,
            {
                firestore,
            },
        );
        this.timesheets = new RegistrationStore(
            this,
            {
                firestore,
            },
        );
        this.reports = new ReportStore(
            this,
            {
                firestore,
            },
        );
        this.dashboard = new DashboardStore(
            this,
            {
                firestore,
            },
        );
        this.projects = new ProjectStore(
            this,
            {
                firestore,
            },
        );
        this.favorites = new FavoriteStore(
            this,
            {
                firestore,
            },
        );
        this.divisions = new DivisionStore(
            this,
            {
                firestore,
                httpsCallable,
            },
        );
        this.router = new RouterStore<IRootStore>(this);
    }

    public dispose() {
        this.config.dispose();
        // this.dashboard.dispose();
        // this.favorites.dispose();
        // this.projects.dispose();
        // this.reports.dispose();
        this.timesheets.dispose();
        this.user.dispose();
        // this.view.dispose();
    }
};
