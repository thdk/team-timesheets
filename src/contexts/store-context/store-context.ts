import { createContext } from 'react';
import { IRootStore } from "../../stores/root-store";

export const StoreContext = createContext<IRootStore>({} as IRootStore);
export const StoreProvider = StoreContext.Provider;
