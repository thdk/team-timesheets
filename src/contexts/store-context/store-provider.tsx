import React from "react";
import { useFirebase } from "../firebase-context/firebase-context";
import { StoreContext } from "./store-context";
import { Store, IRootStore } from "../../stores/root-store";
import { getAuth } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import { getFunctions, httpsCallable } from "firebase/functions";



export const StoreProvider = ({ children, testStore }: React.PropsWithChildren<{ testStore?: IRootStore }>) => {

    const firebase = useFirebase();

    if (testStore) {
        return (
            <StoreContext.Provider value={testStore}>
                {children}
            </StoreContext.Provider>
        );
    }

    const auth = getAuth(firebase);
    const firestore = getFirestore(firebase);
    const functions = getFunctions(firebase);

    const store = (window as any)["store"] = testStore || new Store({
        auth,
        firestore,
        httpsCallable: (name) => httpsCallable(functions, name),
    });

    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    );
}
