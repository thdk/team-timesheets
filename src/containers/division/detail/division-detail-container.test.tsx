import React from "react";
import fs from "fs";
import path from "path";
import type firebase from "firebase";

import { DivisionDetail } from "./";
import { render, waitFor, act, fireEvent } from "@testing-library/react";
import { initializeTestApp, loadFirestoreRules, clearFirestoreData } from "@firebase/rules-unit-testing";
import { Store } from "../../../stores/root-store";
import { useStore } from "../../../contexts/store-context";

const projectId = "division-detail-container";
const app = initializeTestApp({
    projectId,
});

jest.mock("../../../contexts/store-context");

beforeAll(async () => {
    await loadFirestoreRules({
        projectId,
        rules: fs.readFileSync(path.resolve(__dirname, "../../../../firestore.rules.test"), "utf8"),
    });
});

let store: Store;
beforeEach(async () => {
    store = new Store({
        firestore: app.firestore(),
    });

    await store.user.usersCollection.addAsync(
        {
            name: "user 1",
            team: "team-1",
            roles: {
                user: true,
            },
            divisionId: "",
            recentProjects: [],
            tasks: new Map(),
            uid: "user-1",
        },
        "user-1",
    );

    store.auth.setUser({
        uid: "user-1",
        displayName: "user 1",
        email: "email@email.com",
    } as firebase.User);

    await store.divisions.addDocument({
        id: "div-1",
        name: "Division 1",
        icon: "business",
        createdBy: "user-1",
    }, "div-1");

    (useStore as jest.Mock<ReturnType<typeof useStore>>).mockReturnValue(store);
});

afterEach(async () => {
    store.dispose();
    await clearFirestoreData({ projectId });
});

afterAll(() => app.delete());

describe("DivisionDetailContainer", () => {
    it("should not render when there is no active division", () => {
        const {
            asFragment,
        } = render(
            <DivisionDetail />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should display data of active division", async () => {
        await waitFor(() => expect(store.divisions.collection.isFetched).toBe(true));
        await store.divisions.setActiveDocumentId("div-1");

        const {
            getByText,
            container,
            unmount,
        } = render(
            <DivisionDetail />
        );

        await waitFor(() => getByText("business"));

        expect(
            container.querySelector<HTMLInputElement>(".division-detail__name input")?.value
        ).toBe("Division 1");

        unmount();
    });

    it("should update name of active document when name is changed", async () => {
        await waitFor(() => expect(store.divisions.collection.isFetched).toBe(true));
        await store.divisions.setActiveDocumentId("div-1");

        const {
            container,
            getByText,
            unmount,
        } = render(
            <DivisionDetail />
        );

        await waitFor(() => getByText("Name"));

        act(() => {
            fireEvent.change(container.querySelectorAll("input")[0], {
                target: {
                    value: "Division 2",
                },
            });
        });

        expect(store.divisions.activeDocument!.name).toBe("Division 2");

        unmount();
    });
});