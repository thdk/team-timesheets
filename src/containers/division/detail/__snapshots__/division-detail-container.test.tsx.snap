// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DivisionDetailContainer should display data of active division 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <div
      class="flex row horizontal"
    >
      <div
        class="first formField"
      >
        <label
          aria-labelledby="textfield-Name-label"
          class="division-detail__name mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            value="Division 1"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="textfield-Name-label"
                style="transition-duration: 0s;"
              >
                Name
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
      <div
        class="formField"
      >
        <label
          aria-labelledby="textfield-Icon-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--with-leading-icon mdc-text-field--with-trailing-icon"
        >
          <i
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--leading"
          >
            business
          </i>
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            value="business"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
            notch="0"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label mdc-floating-label--float-above"
                id="textfield-Icon-label"
                style="transition-duration: 0s;"
              >
                Icon
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
          <i
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--trailing"
            tabindex="0"
          >
            edit
          </i>
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DivisionDetailContainer should not render when there is no active division 1`] = `<DocumentFragment />`;
